src/txrequest.cpp

```diff
@@ -0,0 +1,685 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <txrequest.h>
+
+#include <crypto/siphash.h>
+#include <net.h>
+#include <primitives/transaction.h>
+#include <random.h>
+#include <uint256.h>
+#include <util/memory.h>
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+
+#include <chrono>
+#include <unordered_map>
+#include <utility>
+
+#include <assert.h>
+
+namespace {
+
+/** The various states a (txhash,peer) pair can be in.
+ *
+ * Note that CANDIDATE is split up into 3 substates (DELAYED, BEST, READY), allowing more efficient implementation.
+ * Also note that the sorting order of EntryTxHash relies on the specific order of values in this enum.
+ *
+ * Expected behaviour is:
+ *   - When first announced by a peer, the state is CANDIDATE_DELAYED until reqtime is reached.
+ *   - Announcemnets that have reached their reqtime but not been requested will be either CANDIDATE_READY or
+ *     CANDIDATE_BEST
+ *   - When requested, an announcement will be in state REQUESTED until expiry is reached.
+ *   - If expiry is reached, or the peer replies to the request (either with NOTFOUND or the tx), the state becomes
+ *     COMPLETED
+ */
+enum class State : uint8_t {
+    /** A CANDIDATE entry whose reqtime is in the future. */
+    CANDIDATE_DELAYED,
+    /** The best CANDIDATE for a given txhash; only if there is no REQUESTED entry already for that txhash.
+     *  The CANDIDATE_BEST is the lowest-priority entry among all CANDIDATE_READY (and _BEST) ones for that txhash. */
+    CANDIDATE_BEST,
+    /** A REQUESTED entry. */
+    REQUESTED,
+    /** A CANDIDATE entry that's not CANDIDATE_DELAYED or CANDIDATE_BEST. */
+    CANDIDATE_READY,
+    /** A COMPLETED entry. */
+    COMPLETED,
+
+    /** An invalid State value that's larger than all valid ones. */
+    TOO_LARGE,
+};
+
+/** An announcement entry. This is the data we track for each txid or wtxid that is announced to us. */
+struct Entry {
+    /** Txid or wtxid that was announced. */
+    const uint256 m_txhash;
+    /** For CANDIDATE_{DELAYED,BEST,READY} the reqtime; for REQUESTED the expiry. */
+    std::chrono::microseconds m_time;
+    /** What peer the request was from. */
+    const uint64_t m_peer;
```

Comment:
The only reason not to is avoiding a dependency on net.h.

Reply:
txrequest.cpp is already including net.h

Reply:
You mean in the header file? Could you make the public functions take `int64_t` and use `NodeId` internally in the cpp file (which already includes net.h)?

Reply:
I think the observable "c++ file dependencies" are only an approximation for what actually matters in terms of code organization: "conceptual dependencies between modules". What you're suggesting is just as much a dependency of txrequest on net in my view - just slightly more hidden.

Reply:
All of this to say: I'd like to avoid a dependency on net, and am therefore using uint64_t as opaque peer identifier rather than NodeId.

But if we feel that's not worth the cost, it should be changed to use NodeId everywhere (in both the .h and the .cpp).

Reply:
I do feel like we should use the same type to represent node ids in all components. Perhaps the purest way to do it would be to move the `typedef NodeId int64_t` somewhere outside net.h, but that seems a bit overkill.

Even if you don't use `NodeId`, is there a reason that you're using `uint64_t` rather than `int64_t` like everywhere else?

Reply:
I've bitten the bullet and converted everything to `NodeId`.

I believe it being a uint64_t dates back to a time when I was trying to squeeze out bits by using less than 64 bits for the peer (and signed bitfields are implementation defined, IIRC). This was a bad idea.